File: oz/game/flipguess.py
from enum import Enum


class Player(Enum):
    Chance = 0
    P1 = 1
    P2 = 2


class Action(Enum):
    Left = 1
    Right = 2
    Heads = -1
    Tails = -2


Action.player_actions = [Action.Left, Action.Right]
Action.chance_actions = [Action.Heads, Action.Tails]
Action.chance_probs = [.5, .5]


class FlipGuess:
    Player = Player
    Action = Action

    def __init__(self):
        self.finished = False
        self.heads = None
        self.p1_choice = None
        self.p2_choice = None
        self.player = Player.Chance

    def is_terminal(self):
        return self.finished

    def utility(self, player=Player.P1):
        if player == Player.P1:
            return self._utility()
        else:
            return -self._utility()

    def _utility(self):
        assert self.is_terminal()
        if self.heads and self.p1_choice == Action.Left:
            return 1
        elif self.p1_choice == self.p2_choice:
            return 3
        else:
            return 0

    def infoset(self):
        if self.player == Player.Chance:
            return self.ChanceInfoset()
        else:
            return self.PlayerInfoset(self.player)

    def act(self, a):
        if self.player == Player.Chance:
            if a == Action.Heads:
                self.heads = True
            elif a == Action.Tails:
                self.heads = False
            else:
                raise ValueError('illegal action: "{}"'.format(a))
            self.player = Player.P1

        elif self.player == Player.P1:
            self.p1_choice = a
            self.player = Player.P2
            if self.heads:
                self.finished = True

        elif self.player == Player.P2:
            self.p2_choice = a
            self.finished = True

        else:
            raise ValueError('illegal action: "{}"'.format(a))

    def __rshift__(self, a):
        self.act(a)
        return self

    class PlayerInfoset:
        actions = Action.player_actions

        def __init__(self, player):
            self.player = player

        def __str__(self):
            return self.player.name

    class ChanceInfoset:
        actions = Action.chance_actions
        probs = Action.chance_probs

File: oz/csrc/games/flipguess.cpp
#include <typeinfo>
#include <cassert>

#include "flipguess.h"

namespace oz {

using namespace std;

auto flipguess_t::infoset() const -> oz::infoset_t {
  Expects(player() != CHANCE);
  return make_infoset<infoset_t>(player_);
}

auto flipguess_t::utility(player_t player) const -> value_t {
  assert (is_terminal());
  value_t u = 0;

  if (heads_ && p1_action_ == action_t::Left) {
    u = 1;
  }
  else {
    if (p1_action_ == p2_action_) {
      u = 3;
    }
    else {
      u = 0;
    }
  }

  return relative_utility(player, u);
}

void flipguess_t::act_(flipguess_t::action_t a) {
  if (player_ == CHANCE) {
    if (a == action_t::Heads) {
      heads_ = true;
    }

    player_ = P1;
  }
  else if (player_ == P1) {
    if (heads_) {
      finished_ = true;
    }

    p1_action_ = a;
    player_ = P2;
  }
  else if (player_ == P2) {
    p2_action_ = a;
    finished_ = true;
  }
}

auto flipguess_t::infoset_t::actions() const -> vector<oz::action_t> {
  static const vector<oz::action_t> player_actions {
      make_action(action_t::Left),
      make_action(action_t::Right),
  };

  return player_actions;
}

auto flipguess_t::infoset_t::str() const -> string {
  if (player_ == CHANCE) {
    return "Chance";
  }
  else if (player_ == P1) {
    return "P1";
  }
  else if (player_ == P2) {
    return "P2";
  }
  else {
    return "?";
  }
}

bool flipguess_t::infoset_t::is_equal(const oz::infoset_t::concept_t& that)
const {
  if (typeid(*this) == typeid(that)) {
    auto that_ = static_cast<const flipguess_t::infoset_t&>(that);
    return player_ == that_.player_;
  }
  else {
    return false;
  }
}

size_t flipguess_t::infoset_t::hash() const {
  return std::hash<player_t>()(player_);
}

auto flipguess_t::chance_actions() const -> map<oz::action_t, prob_t> {
  Expects(player() == CHANCE);

  static const map<oz::action_t, prob_t> actions {
    { make_action(action_t::Heads), 0.5 },
    { make_action(action_t::Tails), 0.5 }
  };

  return actions;
}

} // namespace oz
